File: backend/apps/venues/pictures.py
"""Venue location pictures stored on disk, served at /api/venues/{id}/picture."""

from __future__ import annotations

import base64
import binascii
import re
from pathlib import Path

from django.conf import settings

from .models import Venue

MAX_PICTURE_BYTES = 2 * 1024 * 1024
_DATA_URL = re.compile(
    r"^data:(image/(?:jpeg|jpg|png|webp|gif));base64,(.+)$",
    re.IGNORECASE | re.DOTALL,
)
_EXT_BY_MIME = {
    "image/jpeg": ".jpg",
    "image/jpg": ".jpg",
    "image/png": ".png",
    "image/webp": ".webp",
    "image/gif": ".gif",
}
_MIME_BY_EXT = {
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".png": "image/png",
    ".webp": "image/webp",
    ".gif": "image/gif",
}


def pictures_dir() -> Path:
    root = Path(getattr(settings, "MEDIA_ROOT", Path(settings.BASE_DIR) / "media"))
    dest = root / "venue-pictures"
    dest.mkdir(parents=True, exist_ok=True)
    return dest


def picture_path(venue: Venue) -> Path | None:
    ext = (venue.picture_ext or "").lower()
    if not ext or ext not in _MIME_BY_EXT:
        return None
    return pictures_dir() / f"{venue.id}{ext}"


def picture_url_for(venue: Venue) -> str | None:
    path = picture_path(venue)
    if path is None or not path.is_file():
        return None
    return f"/api/venues/{venue.id}/picture"


def content_type_for(ext: str) -> str:
    return _MIME_BY_EXT.get(ext.lower(), "application/octet-stream")


def decode_picture_data(data: str) -> tuple[bytes, str]:
    """Return (bytes, extension including leading dot) from a data URL."""
    raw = (data or "").strip()
    if not raw:
        raise ValueError("Picture data is empty.")
    match = _DATA_URL.match(raw)
    if not match:
        raise ValueError("Picture must be a PNG, JPEG, WebP, or GIF data URL.")
    mime = match.group(1).lower()
    ext = _EXT_BY_MIME.get(mime)
    if not ext:
        raise ValueError("Unsupported image type.")
    try:
        blob = base64.b64decode(match.group(2), validate=True)
    except (binascii.Error, ValueError) as exc:
        raise ValueError("Picture data is not valid base64.") from exc
    if not blob:
        raise ValueError("Picture data is empty.")
    if len(blob) > MAX_PICTURE_BYTES:
        raise ValueError("Picture must be 2 MB or smaller.")
    return blob, ext


def save_picture(venue: Venue, data: str) -> None:
    blob, ext = decode_picture_data(data)
    dest_dir = pictures_dir()
    for old in dest_dir.glob(f"{venue.id}.*"):
        old.unlink(missing_ok=True)
    (dest_dir / f"{venue.id}{ext}").write_bytes(blob)
    venue.picture_ext = ext
    venue.save(update_fields=["picture_ext", "updated_at"])

File: api/pictures.js
/** Venue location pictures stored under DATA_DIR/venue-pictures/. */

const fs = require("fs");
const path = require("path");

const MAX_PICTURE_BYTES = 2 * 1024 * 1024;
const DATA_URL = /^data:(image\/(?:jpeg|jpg|png|webp|gif));base64,(.+)$/i;
const EXT_BY_MIME = {
  "image/jpeg": ".jpg",
  "image/jpg": ".jpg",
  "image/png": ".png",
  "image/webp": ".webp",
  "image/gif": ".gif",
};
const MIME_BY_EXT = {
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".png": "image/png",
  ".webp": "image/webp",
  ".gif": "image/gif",
};

function picturesDir() {
  const dataDir = process.env.DATA_DIR || path.join(__dirname, "..", "data");
  const dir = path.join(dataDir, "venue-pictures");
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function pictureFile(id, ext) {
  if (!ext) return null;
  const safe = String(ext).toLowerCase();
  if (!MIME_BY_EXT[safe]) return null;
  return path.join(picturesDir(), `${Number(id)}${safe}`);
}

function pictureUrl(row) {
  if (!row || !row.picture_ext) return null;
  const file = pictureFile(row.id, row.picture_ext);
  if (!file || !fs.existsSync(file)) return null;
  return `/api/venues/${row.id}/picture`;
}

function contentTypeFor(ext) {
  return MIME_BY_EXT[String(ext || "").toLowerCase()] || "application/octet-stream";
}

function decodePictureData(data) {
  const raw = String(data || "").trim();
  if (!raw) {
    const err = new Error("Picture data is empty.");
    err.status = 400;
    throw err;
  }
  const match = DATA_URL.exec(raw);
  if (!match) {
    const err = new Error("Picture must be a PNG, JPEG, WebP, or GIF data URL.");
    err.status = 400;
    throw err;
  }
  const ext = EXT_BY_MIME[match[1].toLowerCase()];
  if (!ext) {
    const err = new Error("Unsupported image type.");
    err.status = 400;
    throw err;
  }
  let blob;
  try {
    blob = Buffer.from(match[2], "base64");
  } catch {
    const err = new Error("Picture data is not valid base64.");
    err.status = 400;
    throw err;
  }
  if (!blob.length) {
    const err = new Error("Picture data is empty.");
    err.status = 400;
    throw err;
  }
  if (blob.length > MAX_PICTURE_BYTES) {
    const err = new Error("Picture must be 2 MB or smaller.");
    err.status = 400;
    throw err;
  }
  return { blob, ext };
}

function savePicture(id, data) {
  const { blob, ext } = decodePictureData(data);
  const dir = picturesDir();
  const prefix = `${Number(id)}.`;
  for (const name of fs.readdirSync(dir)) {
    if (name.startsWith(prefix) || name === `${Number(id)}${ext}`) {
      fs.unlinkSync(path.join(dir, name));
    }
  }
  fs.writeFileSync(path.join(dir, `${Number(id)}${ext}`), blob);
  return ext;
}

function readPicture(id, ext) {
  const file = pictureFile(id, ext);
  if (!file || !fs.existsSync(file)) return null;
  return { buffer: fs.readFileSync(file), contentType: contentTypeFor(ext) };
}

module.exports = {
  MAX_PICTURE_BYTES,
  picturesDir,
  pictureUrl,
  savePicture,
  readPicture,
  contentTypeFor,
};
